server/aesdsocket.c

```diff
@@ -0,0 +1,428 @@
+    cleanup();
+    return -1;
+  }
+  freeaddrinfo(server_addr); // no longer needed as we have sockfd 
+  server_addr = NULL;
+
+  if (daemonize_flag) {
+    // ignore signals
+    signal(SIGCHLD, SIG_IGN);
+    signal(SIGHUP, SIG_IGN);
+    pid_t pid = fork();
+    if (pid < 0) {
+      cleanup();
+      return -1;
+    }
+    if (pid > 0) {
+      // parent exits
+      exit(EXIT_SUCCESS); 
+    }
+
+    // create a new session and set process group ID
+    if (setsid() == -1) {
+      perror("setsid()");
+    }
+
+    // change working directory to root
+    chdir("/");
+
+    // redirect STDIOs to /dev/null
+    int devnull = open("/dev/null", O_RDWR);
+    dup2(devnull, STDIN_FILENO);
+    dup2(devnull, STDOUT_FILENO);
+    dup2(devnull, STDERR_FILENO);
+    close(devnull);
+  }
+
+  // listen on the socket
+  rc = listen(sockfd, 5); // listen, allow up to 5 connections to queue
+  if (rc == -1)  {
+    LOG(LOG_ERR, "listen returned -1"); perror("listen()");
+    cleanup();
+    return -1;
+  } 
+
+  while(1) 
+  {
+    int ret = finite_state_machine();
+    if (ret == -1) {
+      cleanup();
+      return -1;
+    }
+  }
+
+  cleanup();
+  return 0; // should never return
+
+} // end main
+
+
+typedef enum states {
+  ACCEPTING_CONNECTIONS,
+  READ_PACKET,
+  WRITE_FILE_SOCKET_ECHO,
+  NUM_STATES
+} states_t;
+
+int finite_state_machine() 
+{
+  static states_t state = ACCEPTING_CONNECTIONS;
+  struct sockaddr_in peer_addr;
+  socklen_t peer_addr_size = sizeof(peer_addr);
+
+  switch(state)
+  {
+    case ACCEPTING_CONNECTIONS:
+    {  
+      //LOG(LOG_INFO, "state: ACCEPTING_CONNECTIONS");
+      // store file descriptor of accepted connection 
+      peerfd = accept(sockfd, (struct sockaddr *) &peer_addr, &peer_addr_size);
+      if (peerfd == -1) {
+        LOG(LOG_ERR, "accept returned -1"); perror("accept");
+        return -1;
+      } else {
+        char peer_addr_str[INET_ADDRSTRLEN];
+        get_ip_str(&peer_addr, peer_addr_str);
+        LOG(LOG_INFO, "Accepted connection from %s", peer_addr_str);
+        // reset recv_buf
+        recv_buf = realloc(recv_buf, 0);
+        recv_buf_size = 0;
+        state = READ_PACKET;
+      } 
+      break;
+
+    }
+
+    case READ_PACKET:
+    { 
+      //LOG(LOG_INFO, "state: READ_PACKET");
+      int recv_temp_size = 256;
+      char recv_temp[recv_temp_size];
+
+      // read from socket until '\n' 
+      int ret = recv(peerfd, &recv_temp[0], recv_temp_size, 0);
+      //LOG(LOG_INFO, "recv returned %d bytes", ret);
+      if (ret == -1 && errno != EINTR) {
+        LOG(LOG_ERR, "recv returned -1"); perror("recv()");
+        free(recv_buf); recv_buf_size = 0; recv_buf = NULL;
+        return -1;
+      } else if (ret == 0) { // end of file (peer socket shutdown)
+        LOG(LOG_INFO, "Peer socket shutdown");
+        state = ACCEPTING_CONNECTIONS; 
+        break;
+      } else if (ret > 0) {
+        recv_buf = realloc(recv_buf, recv_buf_size + ret);
```

Comment:
@jmichael16 what assumption is made here about the size of RAM heap relative to the total size of all packets sent?

Reply:
The heap only gets as big as the single packet of interest. You can see it is reallocd back to zero after writing to the file.

Reply:
It is working in this case but what if you're getting a packet size > 1GB and your heap is of size 800MB?
@dwalkes Can you please review this code? he is doing realloc till one packet size but may fail if packet size is more than heap allocated. Am i right?

Reply:
According to the A5 instructions (Part e): "You may assume the length of the packet will be shorter than the available heap size."

Reply:
okay but what if realloc fails? 🤔 error check is missing